### PR TITLE
planner/cascades: add transformation rule PushSelDownUnionAll

### PR DESCRIPTION
**Pull request by gauss1314**
Signed-off-by: gauss <gauss1314@gmail.com>

### What problem does this PR solve?

Part of #13709.
Add Transformation Rule for PushSelDownUnionAll.

### Check List 

Tests 
 - Unit test

### CODE REVIEW

File: planner/cascades/testdata/transformation_rules_suite_in.json [hunk 1]
```diff
@@ -10,7 +10,8 @@
       "select a, max(b) from t group by a having a > 1",
       "select a, avg(b) from t group by a having a > 1 and max(b) > 10",
       "select t1.a, t1.b, t2.b from t t1, t t2 where t1.a = t2.a and t2.b = t1.b and t1.a > 10 and t2.b > 10 and t1.a > t2.b",
-      "select t1.a, t1.b from t t1, t t2 where t1.a = t2.a and t1.a = 10 and t2.a = 5"
+      "select t1.a, t1.b from t t1, t t2 where t1.a = t2.a and t1.a = 10 and t2.a = 5",
+      "select a, b from ((select a, b from t where a > 1) union all (select a, b from t where b > 5)) as t1"
```
Comment (francis0407) [inline planner/cascades/testdata/transformation_rules_suite_in.json]: This test case won't trigger the rule `PushSelDownUnionAll`. Try `select a, b from ((select a, b from t) union all(select c as a, d as b from t)) as t1 where a > 1`

File: planner/cascades/transformation_rules.go [hunk 2]
```diff
@@ -599,6 +600,40 @@ func (r *PushSelDownJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.Grou
...
+
+	newSelExpr := memo.NewGroupExpr(sel)
+	newSelExpr.Children = append(newSelExpr.Children, childGroup)
+	newSelGroup := memo.NewGroupWithSchema(newSelExpr, childGroup.Prop.Schema)
+
+	newUnionAllExpr := memo.NewGroupExpr(unionAll)
+	newUnionAllExpr.Children = append(newUnionAllExpr.Children, newSelGroup)
+
+	return []*memo.GroupExpr{newUnionAllExpr}, true, false, nil
+}
```
Comment (francis0407) [inline planner/cascades/transformation_rules.go]: UnionAll has more than one child. You should push the selection down to all of it child. For example:
`Selection -> UnionAll -> (Child0, Child1, Child2)` to `UnionAll -> (Selection -> Child0, Selection -> Child1, Selection -> Child2)`

Comment (francis0407) [reply planner/cascades/transformation_rules.go]: `OnTransform` should look like:
```
for i := range oldChildGroups {
    create newSelExpr
    newSelExpr.Child = oldChildGroups[i]
    create newChildGroup from newSelExpr
    newChildGroups[i] = newChildGroup
}
create newUnionAllExpr
set newUnionAllExpr's children as newChildGroups
return newUnionAllExpr
```

**Comment by winoros** [comment]
@gauss1314 You can take a look at `explain select * from t union all select * from t union all from t`. In this case, the `UnionAll` opearator has three children.
So if you want to push the selection down to union. You need push down the selection to each of the union's children.

File: planner/cascades/transformation_rules.go (hunk 2 above)
Comment (gauss1314) [reply planner/cascades/transformation_rules.go]: Thank you, how about this:

    sel := old.GetExpr().ExprNode.(*plannercore.LogicalSelection)
	unionAll := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalUnionAll)
	childGroups := old.Children[0].GetExpr().Children

	newUnionAllExpr := memo.NewGroupExpr(unionAll)
	for _, group := range childGroups {
		newSelExpr := memo.NewGroupExpr(sel)
		newSelExpr.Children = append(newSelExpr.Children, group)
		newSelGroup := memo.NewGroupWithSchema(newSelExpr, group.Prop.Schema)

		newUnionAllExpr.Children = append(newUnionAllExpr.Children, newSelGroup)
	}
	return []*memo.GroupExpr{newUnionAllExpr}, true, false, nil

Comment (francis0407) [reply planner/cascades/transformation_rules.go]: LGTM

**Pull request merged**